zenith_utils/src/pq_proto.rs

```diff
@@ -798,3 +813,156 @@ impl<'a> BeMessage<'a> {
         Ok(())
     }
 }
+
+// Zenith extension of postgres replication protocol
+// See ZENITH_STATUS_UPDATE_TAG_BYTE
+#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
+pub struct ZenithFeedback {
+    // Last known size of the timeline. Used to enforce timeline size limit.
+    pub current_instance_size: u64,
+    // Parts of StandbyStatusUpdate we resend to compute via safekeeper
+    pub ps_writelsn: u64,
+    pub ps_applylsn: u64,
+    pub ps_flushlsn: u64,
+    pub ps_replytime: SystemTime,
+}
+pub const ZENITH_FEEDBACK_FIELDS_NUMBER: u8 = 5;
+
+impl ZenithFeedback {
+    pub fn empty() -> ZenithFeedback {
+        ZenithFeedback {
+            current_instance_size: 0,
+            ps_writelsn: 0,
+            ps_applylsn: 0,
+            ps_flushlsn: 0,
+            ps_replytime: SystemTime::now(),
+        }
+    }
+
+    // use custom serialization to support protocol extensibility
+    pub fn serialize(&self, buf: &mut BytesMut) -> Result<()> {
+        buf.put_u8(ZENITH_FEEDBACK_FIELDS_NUMBER); // # of keys
+        write_cstr(&Bytes::from("current_instance_size"), buf)?;
+        buf.put_i32(8);
+        buf.put_u64(self.current_instance_size);
+
+        write_cstr(&Bytes::from("ps_writelsn"), buf)?;
+        buf.put_i32(8);
+        buf.put_u64(self.ps_writelsn);
+        write_cstr(&Bytes::from("ps_flushlsn"), buf)?;
+        buf.put_i32(8);
+        buf.put_u64(self.ps_flushlsn);
+        write_cstr(&Bytes::from("ps_applylsn"), buf)?;
+        buf.put_i32(8);
+        buf.put_u64(self.ps_applylsn);
+
+        let timestamp = match self.ps_replytime.duration_since(*PG_EPOCH) {
+            Ok(d) => d.as_micros() as i64,
+            Err(e) => -(e.duration().as_micros() as i64),
+        };
+
+        write_cstr(&Bytes::from("ps_replytime"), buf)?;
+        buf.put_i32(8);
+        buf.put_i64(timestamp);
+        Ok(())
+    }
+
+    pub fn parse(mut buf: Bytes) -> ZenithFeedback {
+        let mut zf = ZenithFeedback::empty();
+        let nfields = buf.get_u8();
+        let mut i = 0;
+        while i < nfields {
+            i += 1;
+            let key_cstr = read_null_terminated(&mut buf).unwrap();
+            let key = cstr_to_str(&key_cstr).unwrap();
+            match key {
```

Comment:
Does it make sense to validate that all values have been successfully initialized with correct values? It might cause issues if some of the fields are missing and then someone assumes that all of them were initialized and have non zero values.

Reply:
Well, there's not much to validate.

Feedback messages also serve as keepalive responses, so walsender may send empty values before something was actually replicated. Postgres walsender fine with receiving invalid lsns in a feedback too:
```
/*
	 * The xlog locations that have been written, flushed, and applied by
	 * standby-side. These may be invalid if the standby-side has not offered
	 * values yet.
	 */
```

If some fields are missing, it may be just a different version of the protocol.
Do you think it's worth adding an API version number to the message to control this?